pkg/keycloak/keycloak.go

```diff
@@ -436,51 +453,94 @@ func (h *Handler) reconcileClients(kc *v1alpha1.Keycloak, kcClient KeycloakInter
+		identityProvider := &specRealm.IdentityProviders[i]
+		identityProviderPairsList[identityProvider.Alias] = &v1alpha1.KeycloakIdentityProviderPair{
+			SpecIdentityProvider: identityProvider,
+			KcIdentityProvider:   kcIdentityProviders[identityProvider.Alias],
+		}
+		delete(kcIdentityProviders, identityProvider.Alias)
+	}
+
+	for i := range kcIdentityProviders {
+		identityProvider := kcIdentityProviders[i]
+		identityProviderPairsList[identityProvider.Alias] = &v1alpha1.KeycloakIdentityProviderPair{
+			KcIdentityProvider:   identityProvider,
+			SpecIdentityProvider: nil,
+		}
+	}
 
-			h.reconcileClients(kc, kcClient, objRealm)
-			// h.reconcileUsers(kc, kcClient, kcRealm)
-			// h.reconcileIdentityProviders(kc, kcClient, kcRealm)
+	for i := range identityProviderPairsList {
+		err := h.reconcileIdentityProvider(identityProviderPairsList[i].KcIdentityProvider, identityProviderPairsList[i].SpecIdentityProvider, specRealm.Realm, kcClient)
+		if err != nil {
+			return err
 		}
 	}
 
 	return nil
 }
 
-func (h *Handler) reconcileClient(kcClient, objClient *v1alpha1.KeycloakClient, realmName string, authenticatedClient KeycloakInterface) error {
-	if objClient == nil && !isDefaultClient(kcClient.ClientID) {
-		logrus.Debugf("Deleting client %s in realm: %s", kcClient.ClientID, realmName)
-		err := authenticatedClient.DeleteClient(kcClient.ID, realmName)
+func (h *Handler) reconcileIdentityProvider(kcIdentityProvider, specIdentityProvider *v1alpha1.KeycloakIdentityProvider, realmName string, authenticatedClient KeycloakInterface) error {
+	if specIdentityProvider == nil {
+		logrus.Debugf("Deleting identity provider %s in realm: %s", kcIdentityProvider.Alias, realmName)
+		err := authenticatedClient.DeleteIdentityProvider(kcIdentityProvider.Alias, realmName)
 		if err != nil {
 			return err
 		}
-	} else if kcClient == nil {
-		logrus.Debugf("Creating client %s in realm: %s", objClient.ClientID, realmName)
-		err := authenticatedClient.CreateClient(objClient, realmName)
+	} else if kcIdentityProvider == nil {
+		logrus.Debugf("Creating identity provider %s in realm: %s", specIdentityProvider.Alias, realmName)
+		err := authenticatedClient.CreateIdentityProvider(specIdentityProvider, realmName)
 		if err != nil {
 			return err
 		}
 	} else {
-		if !reflect.DeepEqual(kcClient, objClient) && !isDefaultClient(kcClient.ClientID) {
-			logrus.Debugf("Updating client %s in realm: %s", kcClient.ClientID, realmName)
-			err := authenticatedClient.UpdateClient(kcClient, objClient, realmName)
+		//The API hides the secret value so in order to stop in never being equal we just set it to the spec version
+		kcIdentityProvider.Config["clientSecret"] = specIdentityProvider.Config["clientSecret"]
+		//Ensure the internalID is set on the spec object, this is required for update requests to succeed
+		specIdentityProvider.InternalID = kcIdentityProvider.InternalID
+		if !reflect.DeepEqual(kcIdentityProvider, specIdentityProvider) {
```

Comment:
Manually adding in required fields from kc object to the spec object.

Ideally we would just make sure all missing fields were saved in the spec object from the kc object, but hit this issue trying to use the "mergo" lib https://github.com/imdario/mergo/issues/24

Will look into this and deal with it in a separate PR.